<launch>
    <!-- 基本参数 -->
    <arg name="localization"      default="true"/>
    <arg name="wait_for_transform"  default="0.2"/> 
    <arg name="database_path"     default="~/.ros/rtabmap.db"/>

    <arg     if="$(arg localization)" name="args"  default=""/>
    <arg unless="$(arg localization)" name="args"  default="--delete_db_on_start"/> <!-- 未设置定位时，默认在启动时自动删除~/.ros/rtabmap.db地图文件 -->

    <arg name="rgb_topic" default="/depth_camera/rgb/image_raw"/>
    <arg name="depth_topic" default="/depth_camera/depth/image_raw"/>
    <arg name="camera_info_topic" default="/depth_camera/depth/camera_info"/>
    <!-- <arg name="scan_topic" default="/velodyne_points/scan"/> -->
    
    <arg name="approx_sync" default="false"/>  <!-- 不同步时设为false -->
    <arg name="rviz" default="false"/>

    <include file="$(find car_model)/launch/car_gazebo.launch">
        <arg name="car_model_rviz" value="false" />
    </include>

    <include file = "$(find fast_lio)/launch/mapping_velodyne.launch">
        <arg name="pcd_save_en" value="false"/>
        <arg name="odom_topic" value="/odom" />
    </include>

    <!-- <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan" output="screen">
        <remap from="cloud_in" to="/velodyne_points"/>
        <remap from="scan" to="/velodyne_points/scan"/>

        <rosparam command="load">
            target_frame: velodyne
            transform_tolerance: 0.01

            min_height: 0.4
            max_height: 1.5

            angle_min: -3.1415926
            angle_max: 3.1415926
            angle_increment: 0.003

            range_min: 0.2
            range_max: 100.0

            use_inf: true
            inf_epsilon: 1.0

            concurrency_level: 1

            scan_time: 0.1
        </rosparam>
    </node> -->

    <!-- RTAB-Map 主节点 -->
    <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="$(arg args)">
        <!-- 输入话题配置 -->
        <remap from="rgb/image" to="$(arg rgb_topic)"/>
        <remap from="depth/image" to="$(arg depth_topic)"/>
        <remap from="rgb/camera_info" to="$(arg camera_info_topic)"/>
        <!-- <remap from="scan"            to="$(arg scan_topic)"/> -->

        <!-- 关键参数 -->
        <param name="subscribe_depth" type="bool" value="true"/>
        <param name="subscribe_scan" type="bool" value="false"/>
        <param name="frame_id" type="string" value="base_footprint"/>  <!-- 机器人基坐标系 -->
        <param name="odom_frame_id" type="string" value="odom"/>       <!-- 里程计坐标系 -->
        <param name="approx_sync" type="bool" value="$(arg approx_sync)"/>
        <param name="queue_size" type="int" value="10"/>
        <param name="database_path" type="string" value="$(arg database_path)"/>
        <param name="Rtabmap/DetectionRate" type="double" value="2.0"/>

        <!-- 性能优化参数 -->
        <param name="Mem/IncrementalMemory" type="string" value="true"/>
        <param name="Mem/InitWMWithAllNodes" type="string" value="false"/>
        <param name="Vis/FeatureType" type="string" value="6"/>  <!-- ORB特征 -->
    </node>

    <node name="map2odom" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 /map /odom"/>

    <!-- 可视化工具选择 -->
    <group unless="$(arg rviz)">
        <node name="rtabmapviz" pkg="rtabmap_ros" type="rtabmapviz" output="screen">
        <remap from="rgb/image" to="$(arg rgb_topic)"/>
        <remap from="depth/image" to="$(arg depth_topic)"/>
        <remap from="rgb/camera_info" to="$(arg camera_info_topic)"/>
        <param name="frame_id" type="string" value="base_footprint"/>
        </node>
    </group>

    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find robot_vslam)/rviz/rtab_slam.rviz"/>
</launch>